syntax = "proto3";

import "protos/recipe_nutrition.proto";

option csharp_namespace = "IntelliCook.RecipeSearch.Client";

message SearchRecipesRequest {
    string username = 1;
    repeated string ingredients = 2;
    optional string extra_terms = 3;
    optional uint32 page = 4;
    optional uint32 per_page = 5;
    optional bool include_detail = 6;
}

message SearchRecipesResponse {
    repeated SearchRecipesRecipe recipes = 1;
}

message SearchRecipesRecipe {
    int32 id = 1;
    string title = 2;
    string description = 3;
    repeated SearchRecipesRecipeIngredient ingredients = 4;
    repeated SearchRecipesMatch matches = 5;
    optional SearchRecipesRecipeDetail detail = 6;
}

enum SearchRecipesMatchField {
    TITLE = 0;
    DESCRIPTION = 1;
    INGREDIENTS = 2;
}

message SearchRecipesRecipeIngredient {
    string name = 1;
    optional float quantity = 2;
    optional string unit = 3;
}

message SearchRecipesMatch {
    SearchRecipesMatchField field = 1;
    repeated string tokens = 2;
    optional int32 index = 3;
}

message SearchRecipesRecipeDetail {
    repeated string directions = 1;
    repeated string tips = 2;
    repeated string utensils = 3;
    RecipeNutrition nutrition = 4;
}